Use dict instead of orddict in erl_lint variable table

josevalim opened the pull request:
Once you have more than a dozen of variables in the same
context, the performance with orddict really starts to
deteriorate. Using dict gives better performance overall.

Furthermore, this patch makes the code uniform in using
vtnew() instead of directly creating new variable tables
using []. This means it should be relatively straight-forward
to migrate to another data-structure if desired in the
future.

PS: I have lately been benchmarking and profiling the
compilation stack for perf improvements. This patch
along side @nox earlier pull request provide considerable
performance improvements in erl_lint (which is usually
the slowest pass). I have also identified issues with
beam_jump which I hope to send a pull request or at
least find out more information soon.

josevalim commented:
Please don't merge this pull request. `dict` makes some cases faster but many other cases slower too. I have also looked into `gb_trees` which gave similar performance results as `dict`. Maps performed well but they will have the same issues as `orddict` unless large maps are soon landing in master.

Therefore I would like some feedback on how to handle this:

1. Wait until large maps land in master for v18

2. Create an erl_lint_vt module that abstract all variable table operations. We could then start with a list and upgrade to dict or a map once we reach certain items

Any other ideas?

fenollp commented:
What about using the process' dictionary (and making sure the process is short-lived) ?
This should be faster than anything else and is not necessarily bad.

josevalim commented:
Ok, I have good news. I was able to make the dict case as fast as orddict by exiting earlier when doing dict:merge/3, dict:filter/2 and dict:fold/3. A lot of the operations in erl_lint return new dictionaries and when merging those dictionaries, because there is not fast exit, we end-up traversing all 16 buckets in the dictionary, even if they are empty. Do it a couple thousand times and we get considerable overhead.

I will push improvements to dict.erl in this pull request so we can proceed. The improvements to dict should improve dict usage throughout the stdlib.

josevalim commented:
I have pushed the dict changes. This is good to merge performance wise.

fenollp commented:
You do a lot of different things in the same PR. Isn't that bad?

josevalim commented:
@fenollp I will be glad to split the dict work into a different PR if desired. Right now, I have no idea if any of those changes are welcome, so I won't "fork" the discussion or do extra work unnecessarily.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

okeuday commented:
@josevalim (re: dict changes) It seems better to me if mk_seg was left as mk_seg/1 (i.e., "mk_seg(16)") to make it clearer that there is a dependency on its return value and the value for ?seg_size.  That way that information isn't lost when a viewer sees the modified source code and it is easier to change seg_size in the future.

josevalim commented:
@okeuday I am fine with any approach. Since I am inlining mk_seg, I decided to remove the argument, otherwise the function clause error would become a case clause error which could cause confusion when the number of segment changes (I don't like inlining functions with arity > 0).

okeuday commented:
@josevalim inlining functions with arity > 0 shouldn't be problematic (an example is in the queue module, https://github.com/erlang/otp/blob/maint/lib/stdlib/src/queue.erl#L474-L496).  You could even use erlang:make_tuple(Size, []) to make it more generic, avoiding the need to change in the future.

josevalim commented:
Yup, I am aware you can inline functions with arity > 0. I personally avoid it because it changes the code semantics.

I thought about `erlang:make_tuple/2` and I decided not to use it because I *assumed* it wouldn't be as fast. However, I did some benchmarking and I don't see any diference with using make_tuple. So I am going to move it to `erlang:make_tuple/2` and make sure it is inlined by the compiler too. Thanks!

josevalim commented:
`make_tuple` version with literal inlining pushed.

bjorng commented on lib/stdlib/src/dict.erl:
```diff
@@ -434,6 +436,8 @@ fold_bucket(F, Acc, [?kv(Key,Val)|Bkt]) ->
 fold_bucket(F, Acc, []) when is_function(F, 3) -> Acc.
 
+map_dict(F, #dict{size=0}) when is_function(F, 2) ->
+    new();
```
Wouldn't it work just as well to simply return the input #dict{} record?

josevalim replied:
Indeed. I will change here and filter_dict too.

josevalim commented:
Map and fold updated.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

bjorng commented:
I compiled NBAP-PDU-Contents.erl with and without your patch. Your version turned out to be slower (roughly 2 s compared to 1 s with the original erl_lint).

You can find the file in lib/asn1/test/asn1_SUITE_data/nbapsystem. Use "erlc NBAP-PDU-Contents.asn" to generate an .erl source file. I did my measurements with the '+time' option for erlc.

josevalim commented:
@bjorng I have rewritten it to use `gb_trees`. For compiling the `NBAP-PDU-Contents` file, it was about the same time in my machine and it is still much better for more complex cases (although not quite as good as dict). Can you please give this one a try and let me know?

I have two questions:

1. This patch is NOT ready. Since gb_trees do not have filter, merge nor fold, I have defined a quick implementation inside `erl_lint`. Question: should we extract it to the gb_trees module or leave them in `erl_lint`? If extracting, I would like to propose the following additions:

    * filter/2
    * merge/2 and merge/3
    * foldl/3 and foldr/3 (I am proposing both folds since gb_trees are ordered, I am also fine with a single fold though)

2. In the previous patch, I had some optimizations to Dict which are no longer part of this patch. Although, I could see measurable improvements to compilation. Should I see another pull request with those improvements?

josevalim commented:
Oops, one extra question: 3. Are large maps coming to R18? If so, I could also give them a try.

psyeugenic commented:
@josevalim Large Maps will probably make it in to 18. It depends if we can get it stable. Though for performant fold and filter then maps probably needs iterators. I don't see that making it in.

In other words, don't bank on maps for this PR.

bjorng commented:
1. Extract the functions to the gb_trees module.

2. Yes, please create a pull request with the dict improvements.

I will come back to you when I have re-run the performance tests on the NBAP module.

josevalim commented:
@bjorng Actually, this patch slows down the NBAP module compilation and I don't think it should be included.

I was benchmarking against 17.3 and since my patch contains @nox improvements, I was seeing about the same time because my patch was cancelling @nox perf improvements (i.e. 18 compiles faster than 17.3). If you compare this patch directly against master, it should be slower. Sorry.

I will rewrite this pull request once more to just continue using orddict. I still believe it should be merged since it introduces the more semantically correct vtnew and vtvar functions though.

josevalim commented:
I have pushed the `orddict` version. If the performance is the same and you think it is worth merging, go ahead. If not, just close this. :) Thanks for all the feedback.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

bjorng commented:
We will not merge this patch, as the risks outweigh the benefits. The patch improves the abstraction of the variable tables, but there are still many places in the code that use the variables tables as orddicts or even as ordered lists (lists:merge/1 is called with a list of variable tables).

nox commented:
I'm currently thinking about reworking my refactoring of icrt_export to remove that lists:merge/1 call. I will tell you when I have something.

josevalim commented:
Just a heads up, linting is about 30% to 50% slower on Erlang 18.0-rc2 compared to Erlang 17.0. I have benchmarked both Elixir parser and the ASN file linked here. Since linting is often the slowest step, we may want to take a deeper look at what is happening. I will try to investigate more tomorrow.

psyeugenic commented:
I suspect the new `erl_anno` functions:
Running `erlc +'{eprof,lint_module}' src/erl_parse.erl`:

    FUNCTION                                         CALLS        %     TIME  [uS / CALLS]
    --------                                         -----  -------     ----  [----------]
    ...
    erl_lint:'-nowarn_function/2-lc$^0/1-0-'/2      752700    10.12   646020  [      0.86]
    lists:all/2                                     968552    14.36   916768  [      0.95]
    erl_anno:'-is_string/1-fun-0-'/1                902234    23.58  1505614  [      1.67]
    ---------------------------------------------  -------  -------  -------  [----------]
    Total:                                         4989284  100.00%  6385861  [      1.28]

josevalim commented:
And all this time I was manually adding fprof/eprof to the compiler source. :D Btw, just to clarify, compilation is on average 10% slower on 18.0-rc2 which led me to erl lint. I will play with `erl_anno` now.

josevalim commented:
You are right, this commit is the culprit:

https://github.com/erlang/otp/commit/87a0af476ef82ca2f33d0e15ce324afcfafe3aad

Simply replacing the `is_string/1` check by `is_list/1` in `erl_anno:is_filename/1` almost brings the lint_module compilation to the same times we have seen in 17:

https://github.com/erlang/otp/blob/3c2be863bde4965a1ac61092b1348b96397940cb/lib/stdlib/src/erl_anno.erl#L149-L150

Note that we did not sanity check the value given to the file annotation in the past. I would say checking it is not worth it if it means compilation becomes 10% slower on average. After all, there are many places in the compiler where it will fail if we give it a malformed tree, I wouldn't then special case files.

psyeugenic commented:
The eprof was added to the compiler just recently =)

I removed lists:all/2 yesterday and it seemed to help a bit https://github.com/erlang/otp/compare/master...psyeugenic:egil/opt-erl_anno 
I'm sure more can be done though.

josevalim commented:
Unfortunately the `is_string/1` changes did not help much on my side. :(

After I replaced `is_string/1` by `is_list/1`, 30% of the time for compiling a parser file was on nowarn_bif_clash. I have fixed this in pull request #727. This is a large improvement for compiling parsers, most modules will see just a smaller improvement though.

psyeugenic commented:
Yeah, it was a few percent improvement at best.
Sadly both I but mostly @bjorng spent some energy to improve compile time both in the compiler and in the runtime ..

I'll stop commenting on this thread now.

josevalim commented:
No problem. I will open up pull requests and try to get some feedback. Thanks for all the insights so far!
